### DISCUSSION TITLE
Looking for a workaround for an FnOnce parameter with a non-static lifetime

### DISCUSSION THREAD

**Discussion by pentastich** [asomers/mockall#613] [Questions] [answered]
I have an trait with a function like:
```
    fn eval<'a, T, F>(&self, formula: F) -> Result<T, EvalError>
    where
        F: 'a + FnOnce() -> Result<T, Box<dyn std::error::Error>>;
```
I understand that automock can't handle the *'a* lifetime, and my attempts to use the *mock* macro suggest that approach won't work either.

A few points about my use:
- I don't want to switch to a *static* lifetime because that will force most closures to use *move*, which will complicate use.
- For testing purposes, the only thing that I need to know is that the closure was called, and I could work around that easily enough.

So, my question isn't *can I mock this,* because I'm pretty sure the answer is *no.* My question is whether there's any way I can work around the issue without going to a *static* lifetime.

**Comment by asomers** [comment, marked as answer]
Is this function part of a public trait, part of a trait that you control, or not part of a trait at all?  In the latter two cases, you can conditionally redefine the fuction to use `'static` during `#[cfg(test)]` only.  That's what I do in similar cases.
Another option is to use `mock!` to define a helper function that doesn't have the lifetime, like and then manually define a compliant mock method that wraps the helper.  It's more work, but it will at least allow you to write expectations that your closure was called.  For example:
```rust
mock!{
    pub MyStuct {
        fn __eval<T> -> Result<T, EvalError>;
    }
}
impl MockMyStruct {
    fn eval<'a, T, F>(&self, formula: F) -> Result<T, EvalError>
    where
        F: 'a + FnOnce() -> Result<T, Box<dyn std::error::Error>>
    {
         self.__eval()
    }
}
```

**Reply by pentastich** [reply]
It's in a trait that I control, so I'll give these options a try. The code is from what is basically a spreadsheet evaluation algorithm. I'm implemented it in several languages, and am doing a toy implementation to help me learn Rust.  There's a lot of Rc<RefCell<T>> stuff floating around, including weak pointers, so it's a good way to peek into some language corners!

Thank you!
